### PR TITLE
mgr/cephadm: Add k8s-style event system

### PR DESCRIPTION
**Pull request by sebastian-philipp**
**blocked by**

- [x] #35537

We're getting repeated complains that cephadm is not transparent. We still have not enough visibility of what cephadm is actually doing. 

Adding progress events is going to be complicated, as cephadm contains a declarative state and a loop that tries to make the reality match the configured state. Adding progress events requires some non-trivial bookkeeping, which I'm not willing to do by myself right now. 

Well, this is how kubernetes solve this problem? https://kubernetes.io/docs/tasks/debug-application-cluster/

Does this also work for us? Let's give it a try:

```
$ ceph orch ls --format yaml
events:
- 2020-06-06T23:23:46.910845 service:crash [INFO] "service was created"
placement:
  host_pattern: '*'
service_name: crash
service_type: crash
status:
  container_image_id: 74803e884bea289d2d2d3ebdf6d37cd560499e955595695b1390a89800f4e37a
  container_image_name: docker.io/ceph/daemon-base:latest-master-devel
  created: '2020-06-06T23:23:46.744005'
  last_refresh: '2020-06-06T23:23:57.109881'
  running: 1
  size: 1

$ ceph orch ps --format yaml
container_id: c804e5b26fdd
container_image_id: 74803e884bea289d2d2d3ebdf6d37cd560499e955595695b1390a89800f4e37a
container_image_name: docker.io/ceph/daemon-base:latest-master-devel
created: '2020-06-06T23:23:55.416572'
daemon_id: ubuntu
daemon_type: crash
events:
- 2020-06-06T23:23:55.634997 daemon:crash.ubuntu [INFO] "Deployed crash.ubuntu on
  host 'ubuntu'"
hostname: ubuntu
last_refresh: '2020-06-06T23:44:18.474924'
started: '2020-06-06T23:23:55.503624'
status: 1
status_desc: running
version: 16.0.0-901-g713ef3c

$ ceph orch apply --service_type node-exporter                                                                                                       
Scheduled node-exporter update...

$ ceph orch ls --format yaml                  
events:
- 2020-06-06T23:23:46.910845 service:crash [INFO] "service was created"
placement:
  host_pattern: '*'
service_name: crash
service_type: crash
status:
  container_image_id: 74803e884bea289d2d2d3ebdf6d37cd560499e955595695b1390a89800f4e37a
  container_image_name: docker.io/ceph/daemon-base:latest-master-devel
  created: '2020-06-06T23:23:46.744005'
  last_refresh: '2020-06-06T23:23:57.109881'
  running: 1
  size: 1
---
events:
- 2020-06-06T23:24:10.214343 service:node-exporter [INFO] "service was created"
- '2020-06-06T23:24:10.606714 service:node-exporter [ERROR] "cephadm exited with an
  error code: 1, stderr:INFO:cephadm:Deploy daemon node-exporter.ubuntu ...
  INFO:cephadm:Verifying port 9100 ...
  WARNING:cephadm:Cannot bind to IP 0.0.0.0 port 9100: [Errno 98] Address already
  in use
  ERROR: TCP Port(s) ''9100'' required for node-exporter is already in use"'
placement:
  host_pattern: '*'
service_name: node-exporter
service_type: node-exporter
status:
  running: 0
  size: 1
```

There are obviously many open questions. Some of them:

- [x] Is this a bad idea in general, and we should use HEALTH warnings **instead**?
- [ ] How to convert those events into a meaningful HEALTH warning? Especially clearing the warnings again will be interesting: When exactly is the port conflict solved?
- [x] How to properly present the events in a way that looks good?
- [ ] How to properly manage exceptions in `serve()`
- [x] Is storing lots of events in config-key a bad idea?
- [x] Are we going to overload config-key, if we generate events for all OSDs at once? How to avoid this?
- [ ] This needs testing obviously
- [x] The yaml representation is bad. how to show the events then?
- [x] store the events in memory only. 

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mgfritch** [comment]
This is an interesting idea!

**Comment by jschmid1** [comment]
Indeed interesting!

A couple of random thoughts:

This is super helpful for troubleshooting/debugging and adds a lot of transparency.
The way that kubectl displays it looks great too, we could probably adopt this in any non-yaml/json format.

Otoh, this feels a bit redundant, the 'events' should be present in the logs. (If not, we did a bad job at logging). It should be rather easy to filter out 'events' from the logs and print them upon request. 

Using a separate 'events' implementation would enable us to react on them This however adds a lot of complexity and probably requires a more sophisticated solution.

**Comment by sebastian-philipp** [comment]
> Otoh, this feels a bit redundant, the 'events' should be present in the logs.

Right. This is mainly for diagnosing obvious errors. I'm just keeping 5 events per subject in the system.

> (If not, we did a bad job at logging). It should be rather easy to filter out 'events' from the logs and print them upon request.

We already have 

https://github.com/ceph/ceph/blob/779de8c3d6f291c76010f3cd826cde6f94255e08/src/pybind/mgr/k8sevents/module.py#L159

but I don't see this as particular helpful for finding domain-specific issues. We'd probably have to wring somethinge else.

> Using a separate 'events' implementation would enable us to react on them This however adds a lot of complexity and probably requires a more sophisticated solution.

Adding context information is already pretty hard to do right. No idea if that would also work for a generic solution.

File: src/pybind/mgr/cephadm/inventory.py [hunk 1]
```diff
@@ -427,4 +429,74 @@ def add_daemon(self, host, dd):
...
+
+            # limit to five events for now.
+            self.events[event.kind_subject()] = self.events[event.kind_subject()][-5:]
+        else:
+            self.events[event.kind_subject()] = [event]
+
+        self.mgr.set_store(
+            EVENT_STORE_PREFIX + event.kind_subject(),
+            json.dumps({
+                'evs': [ev.to_json() for ev in self.events[event.kind_subject()]]
```
Comment (votdev) [inline src/pybind/mgr/cephadm/inventory.py:469]: Can we use `events` here to make it more clear?
```suggestion
                'events': [ev.to_json() for ev in self.events[event.kind_subject()]]
```

**Comment by votdev** [comment]
I got the creation message, but not the error which let the deployment fail.
```
# cat <<EOF > /tmp/cephadm-apply.yml
service_type: mon
placement:
  hosts:
    - mon0
EOF
# ceph orch apply -i /tmp/cephadm-apply.yml
# date
Mon 08 Jun 2020 01:26:50 PM UTC
# ceph orch ls --format yaml
events:
- 2020-06-08T13:21:13.623776 service:mon [INFO] "service was created"
placement:
  hosts:
  - hostname: mon0
    name: ''
    network: ''
service_name: mon
service_type: mon
status:
  running: 0
  size: 1
# ceph orch ls    
NAME  RUNNING  REFRESHED  AGE  PLACEMENT  IMAGE NAME  IMAGE ID   
mon       0/1  -          -    mon0       <unknown>   <unknown>
```
The error found in the log file is
```
2020-06-08T13:21:15.174+0000 7f3fcdc8d700 20 mgr ~Gil Destroying new thread state 0x563e8e22ec60
2020-06-08T13:21:15.178+0000 7f3fc413a700  0 [cephadm ERROR root] Failed to apply mon spec ServiceSpec({'pl>
Traceback (most recent call last):
  File "/ceph/src/pybind/mgr/cephadm/module.py", line 1872, in _apply_all_services
    if self._apply_service(spec):
  File "/ceph/src/pybind/mgr/cephadm/module.py", line 1839, in _apply_service
    create_func(daemon_id, host, network)  # type: ignore
  File "/ceph/src/pybind/mgr/cephadm/services/cephadmservice.py", line 91, in create
    raise OrchestratorError('Must set public_network config option or specify a CIDR network, ceph addrvec,>
orchestrator._interface.OrchestratorError: Must set public_network config option or specify a CIDR network,>
2020-06-08T13:21:15.178+0000 7f3fc413a700 20 log_channel(cephadm) update_config log_to_monitors {default=tr>
```
It would be great to see that error in the `events` list.

**Comment by sebastian-philipp** [comment]
> It would be great to see that error in the `events` list.

So, you think we should go ahead with this approach?

**Comment by votdev** [comment]
> > It would be great to see that error in the `events` list.
> 
> So, you think we should go ahead with this approach?

Got another problem while testing my PR to deploy services via Dashboard where your PR was really helpful to identify the problem easily.
```
# ceph orch ls --format yaml
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2020-06-08T14:44:37.004+0000 7f62acc81700 -1 WARNING: all dangerous and experimental features are enabled.
2020-06-08T14:44:37.048+0000 7f62a759e700 -1 WARNING: all dangerous and experimental features are enabled.
events:
- 2020-06-08T14:19:08.879649 service:grafana [INFO] "service was created"
- '2020-06-08T14:19:09.792884 service:grafana [ERROR] "cephadm exited with an error
  code: 1, stderr:INFO:cephadm:Deploy daemon grafana.mgr0 ...
  INFO:cephadm:Verifying port 3000 ...
  WARNING:cephadm:Cannot bind to IP 0.0.0.0 port 3000: [Errno 98] Address already
  in use
  ERROR: TCP Port(s) ''3000'' required for grafana is already in use"'
- '2020-06-08T14:19:10.299231 service:grafana [ERROR] "cephadm exited with an error
  code: 1, stderr:INFO:cephadm:Deploy daemon grafana.mgr0 ...
  INFO:cephadm:Verifying port 3000 ...
  WARNING:cephadm:Cannot bind to IP 0.0.0.0 port 3000: [Errno 98] Address already
  in use
  ERROR: TCP Port(s) ''3000'' required for grafana is already in use"'
- 2020-06-08T14:43:51.040352 service:grafana [INFO] "service was created"
- '2020-06-08T14:43:51.667427 service:grafana [ERROR] "cephadm exited with an error
  code: 1, stderr:INFO:cephadm:Deploy daemon grafana.mgr0 ...
  INFO:cephadm:Verifying port 3000 ...
  WARNING:cephadm:Cannot bind to IP 0.0.0.0 port 3000: [Errno 98] Address already
  in use
  ERROR: TCP Port(s) ''3000'' required for grafana is already in use"'
placement:
  label: aaaa
service_name: grafana
service_type: grafana
status:
  running: 0
  size: 1
```

**Comment by sebastian-philipp** [comment]
@neha-ojha + @jecluis . Are you ok with putting O(daemons) objects into config-key ? Or do I need a different way to store the data?

**Comment by neha-ojha** [comment]
> @neha-ojha + @jecluis . Are you ok with putting O(daemons) objects into config-key ? Or do I need a different way to store the data?

@sebastian-philipp I need to understand the context for this, will revisit it next week.

**Comment by sebastian-philipp** [comment]
ok the context is: cephadm is really opaque and no one knows what cephadm is doing and what went wrong except for looking into the MGR log file. We're talking about things like `deployed daemon x on host y`, `will remove daemon x from host y`. `service x was created`, `failed to deploy daemon x on host y, cause foobar`. With the exception of the last message, this is nothing we can push to the user via HEALTH, which is a different topic. But I need to store those messages for each daemon and right now, I'm storing at max 5 messages. Now the Q: is config-key capable to storing a json document per daemon?

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-06-29_09:26:42-rados:cephadm-wip-swagner-testing-2020-06-26-1524-distro-basic-smithi/

* Failure: https://tracker.ceph.com/issues/46178

**Comment by sebastian-philipp** [comment]
```
2020-06-26 14:24:48,261.261 INFO:__main__:Running ['./bin/ceph', 'mgr', 'self-test', 'module', 'orchestrator']
2020-06-26T14:24:48.439+0000 7f5da5505700 -1 WARNING: all dangerous and experimental features are enabled.
2020-06-26T14:24:48.479+0000 7f5da5505700 -1 WARNING: all dangerous and experimental features are enabled.
Error EPERM: Test failed: Remote method threw exception: Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/src/pybind/mgr/orchestrator/module.py", line 1275, in self_test
    raise_if_exception(e2)
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/src/pybind/mgr/orchestrator/_interface.py", line 651, in raise_if_exception
    raise e
orchestrator._interface.OrchestratorError: hello

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/src/pybind/mgr/orchestrator/module.py", line 1278, in self_test
    assert e.args == ('hello', 'world')
AssertionError

2020-06-26 14:24:48,877.877 INFO:__main__:Running ['./bin/ceph', 'log', 'Ended test tasks.mgr.test_module_selftest.TestModuleSelftest.test_orchestrator']
2020-06-26 14:24:49,674.674 INFO:__main__:test_orchestrator (tasks.mgr.test_module_selftest.TestModuleSelftest) ... ERROR
2020-06-26 14:24:49,674.674 INFO:__main__:Stopped test: test_orchestrator (tasks.mgr.test_module_selftest.TestModuleSelftest) in 27.142287s
```

**Comment by sebastian-philipp** [comment]
enkins test dashboard backend

**Comment by sebastian-philipp** [comment]
jenkins test dashboard backend

File: src/pybind/mgr/cephadm/inventory.py [hunk 2]
```diff
@@ -440,3 +441,56 @@ def daemon_cache_filled(self):
...
+        if event.kind_subject() not in self.events:
+            self.events[event.kind_subject()] = [event]
+
+        for e in self.events[event.kind_subject()]:
+            if e.message == event.message:
+                return
+
+        self.events[event.kind_subject()].append(event)
+
+        # limit to five events for now.
```
Comment (jmolmo) [inline src/pybind/mgr/cephadm/inventory.py:463]: What about to write the purged events to manager log?

Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/inventory.py:463]: everything is written to the logs anyway.

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-07-01_09:26:21-rados:cephadm-wip-swagner-testing-2020-07-01-0956-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
jenkins retest this please

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-07-15_10:40:40-rados:cephadm-wip-swagner2-testing-2020-07-15-1022-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
cephadm/module.py:45: note: In module imported here:
cephadm/inventory.py: note: In member "cleanup" of class "EventStore":
cephadm/inventory.py:500: error: Unpacking a string is disallowed
cephadm/inventory.py:501: error: Cannot determine type of 'k_s'
cephadm/inventory.py:504: error: Cannot determine type of 'k_s'
cephadm/inventory.py:507: error: Cannot determine type of 'k_s'
Found 4 errors in 1 file (checked 11 source files)

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-07-16_14:28:11-rados:cephadm-wip-swagner-testing-2020-07-16-1149-distro-basic-smithi/

**Comment by smithfarm** [comment]
@sebastian-philipp Would it make sense to also emit these events to the cluster log? 

https://docs.ceph.com/docs/master/dev/logging/

**Comment by sebastian-philipp** [comment]
everything is logged in the mgr log right now. do we really need to log there as well?

**Comment by sebastian-philipp** [comment]
image

hehe. I think you clicked on edit, instead of quote :-)

**Comment by smithfarm** [comment]
> hehe. I think you clicked on edit, instead of quote :-)

Yes. That's a bad habit of mine ;-)

**Comment by smithfarm** [comment]
> everything is logged in the mgr log right now. do we really need to log there as well?

Depends on whom you mean by "we". Developers or users?

Searching the individual MGR logs for significant events might be easy for developers (?), but most users probably find it hard and mystifying.

The cluster log's purpose is to aggregate significant cluster-wide events in a single place. In my understanding, these "k8s-style" events would be a good fit for the cluster log.

**Pull request merged**